src/aiokatcp/sensor.py

```diff
@@ -714,3 +715,69 @@ def copy(self) -> Dict[str, Sensor]:
     values.__doc__ = dict.values.__doc__
     items.__doc__ = dict.items.__doc__
     copy.__doc__ = dict.copy.__doc__
+
+
+# I'm ever so slightly nervous about inheritance and having a metaclass.
+class AggregateSensor(Sensor, metaclass=ABCMeta):
+    """A Sensor with its reading determined by several other Sensors.
+
+    This is an abstract class, the user must implement :meth:`_update_aggregate`
+    in order to define the logic of how the aggregate sensor's reading will be
+    determined by those of the sensor set to which it is targeted.
+
+    Parameters are all as per :class:`Sensor`, with the exception of
+    `target`, which is the :class:`SensorSet` from which the aggregated sensor
+    will determine its own reading.
+    """
+
+    def __init__(
+        self,
+        target: SensorSet,
+        sensor_type: Type[_T],
+        name: str,
+        description: str = "",
+        units: str = "",
+        default: _T = None,
+        initial_status: Sensor.Status = Sensor.Status.UNKNOWN,
+        *,
+        status_func: Callable[[_T], Sensor.Status] = _default_status_func,
+        auto_strategy: Optional["SensorSampler.Strategy"] = None,
+        auto_strategy_parameters: Iterable[Any] = (),
+    ) -> None:
+        Sensor.__init__(
+            self,
+            sensor_type=sensor_type,
+            name=name,
+            description=description,
+            units=units,
+            default=default,
+            initial_status=initial_status,
+            status_func=status_func,
+            auto_strategy=auto_strategy,
+            auto_strategy_parameters=auto_strategy_parameters,
+        )
+        self.target = target
+        for sensor in self.target.values():
+            if sensor is not self:  # This sensor might be in the set.
+                sensor.attach(self._update_aggregate)
+
+        self.target.add_add_callback(self._sensor_added)
+        self.target.add_remove_callback(self._sensor_removed)
+
+    @abstractmethod
+    def _update_aggregate(self, updated_sensor: Sensor, reading: Reading):
+        return
+
+    def _sensor_added(self, sensor: Sensor):
+        """Add the update callback to a new sensor in the set."""
+        if sensor is not self:
+            sensor.attach(self._update_aggregate)
+        # The new sensor might actually affect the aggregate reading.
+        self._update_aggregate(sensor, sensor.reading)
```

Comment:
The only thing I'd originally wanted to isolate was the addition (or removal) of the callback. The un-indenting of the next line wasn't by accident, but now I'm questioning whether it's actually necessary.

The addition or removal of the AggregateSensor to/from the set shouldn't affect the value, so I guess an update wouldn't need to be triggered?

Reply:
And if the callback actually uses the sensor reading it's given, you have a cyclic dependency, where adding the aggregate to the set causes its value to be updated based on its previous value, but just once. That seems pretty weird. I think it would be more consistent if the aggregate sensor was completely blind to itself i.e. was never the argument to `_update_aggregate`.